### PR TITLE
[ Hackathon  3rd  No.27 ]  add  paddle.incubate.sparse.any

### PR DESCRIPTION
**Pull request by OccupyMars2030**
### PR types

New features 
### PR changes

APIs 
### Describe

add  paddle.incubate.sparse.any

contest website: https://github.com/PaddlePaddle/Paddle/issues/43938

暂时不用评审本PR，我还没有开始开发，只是先提了PR让我尽快进入开发状态

### CODE REVIEW

**Comment by OccupyMars2030** [comment]
References:
[sparse_csr_tensor](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/sparse/sparse_csr_tensor_cn.html)   
[sparse_coo_tensor](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/sparse/sparse_coo_tensor_cn.html)   
[all new sparse APIs and their corresponding PRs](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/release_note_cn.html#id11)
image

[scipy.sparse.coo_array](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_array.html#scipy-sparse-coo-array)
[scipy.sparse.csr_array](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csr_array.html#scipy.sparse.csr_array)
[implementation of coo, csr in Java](https://programmer.ink/think/implementation-of-coo-and-csr-based-on-array-form-for-sparse-matrix.html)

sparse matrix operation:  
[scipy.sparse.linalg](https://docs.scipy.org/doc/scipy/reference/sparse.linalg.html#module-scipy.sparse.linalg)

[torch sparse](https://pytorch.org/docs/stable/sparse.html)


_**maybe**_ useful material:  
https://github.com/PaddlePaddle/Paddle/releases

**Comment by OccupyMars2025** [comment]
暂时不用评审本PR，我还没有开始开发，只是先提了PR让我尽快进入开发状态

**Comment by OccupyMars2025** [comment]
```python
weight_top_u = lambda x, y: 1.0 - 20.0 * abs(x)
bc_top_u = psci.bc.Dirichlet('u', rhs=1.0, weight=weight_top_u)
bc_top_v = psci.bc.Dirichlet('v', rhs=0.0)
bc_down_u = psci.bc.Dirichlet('u', rhs=0.0)
bc_down_v = psci.bc.Dirichlet('v', rhs=0.0)
bc_left_u = psci.bc.Dirichlet('u', rhs=0.0)
bc_left_v = psci.bc.Dirichlet('v', rhs=0.0)
bc_right_u = psci.bc.Dirichlet('u', rhs=0.0)
bc_right_v = psci.bc.Dirichlet('v', rhs=0.0)

pde.add_bc("top", bc_top_u, bc_top_v)
pde.add_bc("down", bc_down_u, bc_down_v)
pde.add_bc("left", bc_left_u, bc_left_v)
pde.add_bc("right", bc_right_u, bc_right_v)

```

**Comment by OccupyMars2030** [comment]
```C
int csr_row[rows + 1] = { 0 };
for (int i = 0; i < nnz; i++)
    csr_row[coo_row[i] + 1]++;
for (int i = 0; i < rows; i++)
    csr_row[i + 1] += csr_row[i];
```

This is the core  C  code for transforming  a sparse matrix in COO format to one in CSR format.
This method has no considerations for duplicated COO entries.

**Comment by OccupyMars2025** [comment]
image

https://pytorch.org/docs/stable/generated/torch.sparse_coo_tensor.html#torch.sparse_coo_tensor

What's the point of setting  dimension size to  0  ?
What does "dense_dim"  mean ?

**Comment by OccupyMars2030** [comment]
https://pytorch.org/docs/stable/sparse.html#sparse-csr-tensor

 using MKL and MAGMA backends. CUDA support does not exist as of now.

**Comment by OccupyMars2030** [comment]
2022/9/3  18:00  -  9/4  11:40 :
paddle\fluid\pybind\exception.cc  
https://pybind11.readthedocs.io/en/stable/advanced/exceptions.html

**Comment by OccupyMars2030** [comment]
https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/any_cn.html    
Modify typos on this page when I have time.

https://pytorch.org/docs/stable/generated/torch.any.html

**Comment by OccupyMars2030** [comment]
# The reference chain for constructing  SparsCooTensor and  SparseCsrTensor,  how  the C++ method  is  bound  to python,  it is bound  to  a  python module  "core_avx.eager"  or  "core_noavx.eager"  :  
paddle\fluid\pybind\eager_functions.cc : 
image


paddle\fluid\pybind\eager_functions.cc :
image

paddle\fluid\pybind\eager_functions.cc :  
image
image

paddle\fluid\pybind\eager_functions.cc :  
image

paddle\fluid\pybind\eager.cc :  
image
image

paddle\fluid\pybind\pybind.cc :    (it seems that all C++ classes, methods are bound to python  by this file)  
image


python\paddle\fluid\core.py :   Now  it can  be bound  to python  
image

**Comment by OccupyMars2030** [comment]
## [What is monkey patch ?](https://en.wikipedia.org/wiki/Monkey_patch)  
python\paddle\fluid\dygraph\varbase_patch_methods.py  
image

**Comment by OccupyMars2030** [comment]
python\paddle\fluid\dygraph\varbase_patch_methods.py
image
image

paddle\fluid\pybind\eager_method.cc
image

**Comment by OccupyMars2030** [comment]
# but how tensor method   "to_sparse_csr"  is bound to  python end ?  It is bound  to python  class  Tensor as its method.
paddle\fluid\pybind\eager_method.cc 
image

**Comment by OccupyMars2030** [comment]
## paddle::optional
https://www.boost.org/doc/libs/1_80_0/libs/optional/doc/html/index.html
paddle\utils\optional.h
https://blog.csdn.net/zzhongcy/article/details/87623430

**Comment by OccupyMars2030** [comment]
# to_sparse_coo_impl
paddle\phi\api\lib\sparse_api_custom_impl.cc
#  to_sparse_coo_grad,  to_dense_grad
python\paddle\fluid\tests\unittests\test_sparse_utils_op.py
image
# "mask" a tensor with an index tensor  (coo_to_dense_grad has used this)
paddle\phi\kernels\sparse\cpu\mask_kernel.cc
# about the backprop through the sparse tensors: when implementing grad kernel, it seems that all the indices with default zero values in a sparse tensor  are just ignored,  so  when training the neural network,    all default zero values in a sparse tensor  will remain unchanged, is that ok ???
1.  In the coo_to_dense_grad,   out_grad is  masked  by  x.indices()  to  get  x_grad.   But in my opinion,  out_grad should just be returned as x_grad.  It is confusing  for me to understand  backprop when the question is not  expressed in  formal math expressions.
2.  [SparseReluGradKernel](https://github.com/PaddlePaddle/Paddle/pull/40959/files#diff-4b6b0713eaf958d2a21737210a5bf806dfa9ddb4582a1639fffc253da7a0edb1R28)
# parse yaml and generate code ?
paddle\fluid\eager\auto_code_generator\generator\python_c_gen.py
image

**Comment by OccupyMars2030** [comment]
# code generation
paddle\fluid\eager\auto_code_generator\generator\python_c_gen.py    uses python to generate Python C  code. 
where  is the file "python_c_gen.py"  used ?  
paddle\fluid\eager\auto_code_generator\generator\CMakeLists.txt
image
After compiling PaddlePaddle source code on the local machine,  you  can  find out that  this  file  is  generated: 
paddle\fluid\pybind\eager_op_function.cc
image
# code generation related to sparse tensor
paddle\phi\api\yaml\sparse_ops.yaml
paddle\phi\api\yaml\sparse_backward.yaml
paddle\phi\api\yaml\generator\sparse_api_gen.py
paddle\phi\api\yaml\generator\sparse_bw_api_gen.py
paddle\phi\api\lib\CMakeLists.txt
image
image

**Comment by OccupyMars2030** [comment]
# Is there need to  add  grad kernel for  sparse any?   Study  how  paddle.any is implemented. 
# In cpu kernel of paddle.any, it seems to use "any" method of Eigen Tensor  to execute the operation
paddle\phi\kernels\funcs\reduce_functor.h
image
When using this functor,  the "X" is an Eigen Tensor:
unsupported\Eigen\CXX11\src\Tensor\TensorBase.h
image
